### ISSUE TITLE
Consolidate and organize block/style copy/paste menus

### ISSUE THREAD

**Issue by t-hamano** [WordPress/gutenberg#58650]
Related issues/PRs:

- #44418
- #45477
- #57254
- #55611

## What problem does this address?

There are two copy actions in the current block settings dropdown: "Copy" and "Copy style". The two are internally the same and simply copy blocks. Which one to paste depends on whether you select "Paste" or "Paste styles".

It might be possible to modify the current copy style action to separate block copying and style copying, but I think the following problems may occur:

- If you copy a block, you cannot paste the style.
- If you copy the style, you cannot paste the block.

I feel like it would be more flexible and understandable if there was just one copy action, like in Google Sheets, and we could choose whether to paste blocks or styles when pasting.

## What is your proposed solution?

There was discussion in #44418 and #45477 as to whether block-style copying is necessary at all, but I propose to integrate copy actions and group them as follows.

Before: 

```
Copy
Duplicate
Add before
Add after
---
Copy styles
Paste styles
```

After:

```
Copy
Duplicate
Paste
Paste styles
---
Add before
Add after
```

![mockup](https://github.com/WordPress/gutenberg/assets/54422211/fee11d64-21de-471d-8734-7e056879df12)

If this block toolbar dropdown supports nesting in the future, the following UI may be possible.

![mockup-nested](https://github.com/WordPress/gutenberg/assets/54422211/6b86564f-4abb-4e01-9c49-d916ada84b5e)

If it makes sense to accomplish this, I believe that the following secondary tasks are also necessary.

- Add "Paste styles" (`cmd + shift + v`) shortcut in all editor instances
- Add "Paste styles" to the keyboard shortcut modal in all editor instances
- Display shortcut keys for "Copy", "Paste" and "Paste styles" in the block dropdown menu

cc: @kevin940726 , who developed the paste style, and the @WordPress/gutenberg-design team

**Comment by jasmussen** [comment]
Length of the dropdown is definitely the main issue at hand here. We want to avoid getting into situations where you have the dropdown actually have a scrollbar, as much as possible. To that end I'd support primarily your last mockup; a single copy action, and nested paste actions.

**Comment by kevin940726** [comment]
I always support the single copy action from the beginning but there were different concerns 😆. One thing to consider for the nested dropdown is, AFAIK, we don't currently support that in our component library. It might not be trivial to implement considering different UX needs either. However, the adoption of the `ariakit` library should help a lot here! (c.c. @WordPress/gutenberg-components)

**Comment by jasmussen** [comment]
[DropdownMenu V2](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2--docs#with-slot-fill) supports nesting now! There are some style issues if I recall correctly, with the menu as it extends from the block toolbar and needs the dark borders and no drop-shadow.

**Comment by ciampo** [comment]
> [DropdownMenu V2](https://wordpress.github.io/gutenberg/?path=/docs/components-experimental-dropdownmenu-v2--docs#with-slot-fill) supports nesting now! There are some style issues if I recall correctly, with the menu as it extends from the block toolbar and needs the dark borders and no drop-shadow.

Just a word of caution re. the V2 of `DrodownMenu`. While style issues shouldn't be difficult to solve, we still haven't ironed out all the wrinkles of using it in the editor, _particularly_ in the block settings toolbar.

I have a [draft PR](https://github.com/WordPress/gutenberg/pull/57996) where I've been exploring using the new component in the block settings toolbar, but there are still a few blockers that need to be fixed, in particular:
- around the modality of the dropdown and some second-order effects of that decision
- around where in the dom the dropdown contents render (ie. bridging the gap with how the legacy `Popover` renders in a `Popover.Slot`)
- around backward compatibility when `Slot`/`Fill` are involved — ie., if we update the `DropdownMenu` to V2, all the legacy `MenuItem` items injected by third party plugins/etc won't work

We're getting there, but it's a tricky process and in the past months we've been prioritising Data Views, and so we haven't made a ton of progress on this front (cc @mirka )

**Comment by t-hamano** [comment]
Thank you everyone for your advice. Regardless of whether this menu is replaced by `DropdownMenu` V2 in the future, my suggestion will not increase the number of dropdown menu items. Therefore, if you agree with this proposal itself, I would like to work on this issue 👍

**Comment by bacoords** [comment]
I agree with @t-hamano - this suggestion is great because it doesn't increase the number of dropdown items and (as opposed to moving the paste options into yet another deeper layer requiring even more interaction) it makes them feel even more discoverable.

**Comment by richtabor** [comment]
Add before and after are **much more** signifiant to the editing experience than copying/pasting styling. I would not prioritize copy over block insertion.

I do feel a bit of hesitancy on putting the "Copy" control nested one level. It's such an integral part of the experience, I don't know that it's worth nested to save one spot on the options menu. 

I lean towards something less dramatic: 

- Copy
- Duplicate
- Add before
- Add after
- ---------
- Copy/Paste as >
- - Copy as text (a concept, to only copy content)
- - Copy styles
- - Paste styles 
- - Paste to replace (a concept, not something we currently support). 

This is related to https://github.com/WordPress/gutenberg/issues/49271 as well.

**Comment by kevin940726** [comment]
@richtabor I don't think @t-hamano's proposal is to nest the copy styles action. I believe the proposal is to remove that action instead. They currently do the same thing anyway.

**Comment by richtabor** [comment]
We should not remove the explicit copy styles control as then the capability would be undiscoverable.

**Comment by t-hamano** [comment]
> We should not remove the explicit copy styles control as then the capability would be undiscoverable.

My question is whether it is necessary to make it discoverable to be able to copy styles. Isn't it enough to have a choice of what to paste when you paste?

Perhaps there are apps that have menus to copy only specific contexts, but in the Google app, for example, there is only one "copy".

Docs:

![image](https://github.com/WordPress/gutenberg/assets/54422211/499dee67-5b5e-40a3-8f74-656b811b9ea3)

Sheets:

![image](https://github.com/WordPress/gutenberg/assets/54422211/b9c464e0-57c0-41bb-bc07-62fe8f7bab71)

Slides:

![image](https://github.com/WordPress/gutenberg/assets/54422211/6bdbc8af-1ad5-4831-8e51-8ada04b8ffd5)

**Comment by richtabor** [comment]
> My question is whether it is necessary to make it discoverable to be able to copy styles. 

Yes, I do think it is necessary. Otherwise, you have to _know_ you can just copy styles. 

In Google sheets, you can't copy just the styles, you can copy everything and paste without formatting. Copy styles is where you can copy the block style attributes, and paste those attributes on another block. Essentially the opposite of pasting without formatting. It's not an expected behavior without the clarity, even if the controls are technically doing the same thing (copying the block and the styles).
